test_runner/regress/test_tenants.py

```diff
@@ -475,3 +476,37 @@ def only_int(samples: List[Sample]) -> int:
     assert counts
     log.info(f"directory counts: {counts}")
     assert counts[2] > COUNT_AT_LEAST_EXPECTED
+
+
+def test_tenants_parallel_creation(neon_simple_env: NeonEnv):
+    """
+    (Relaxed) regression test for issue that led to https://github.com/neondatabase/neon/pull/9268
+    Create many tenants in parallel and then restart endpoints
+    """
+    env = neon_simple_env
+    pageserver_http_client = env.pageserver.http_client()
+
+    # This param needs to be 200 to reproduce the limit issue
+    n_threads = 8
+    barrier = threading.Barrier(n_threads)
+
+    def test_tenant():
+        # Create new tenant
+        tenant_id, _ = env.create_tenant()
+        endpoint = env.endpoints.create_start("main", tenant_id=tenant_id)
+        time.sleep(2)
+        endpoint.stop()
+        time.sleep(2)
```

Comment:
What do these sleeps do?

Reply:
the point is that there is some period of time where all threads have endpoints running: this should ideally create parallel connections on the storage broker instead of just having many point like connections.

Reply:
the point of the sleeps is to create a period of time for the race condition to work, during which there is multiple running endpoints.

Reply:
Is there a mechanism to wait for endpoints to complete initialization and become fully available? Seems like that, combined with a barrier, would be more reliable than an arbitrary wait.

Reply:
I suppose some way to wait on the log message about the walreceiver being started would work. These waits are mostly a defense in depth though, present so that I don't have to analyze in depth whether there is hidden race conditions that make the test more flaky if they are not present at all (most efficient use of my time; it's just a test after all). So i don't know what to really wait on or not.

The sleep above is more needed. I think if I remove them, the test will not reproduce the issue any more. So personally I think it's fair to have sleeps, although I'm happy to implement concrete proposals for what should replace the sleeps.

Reply:
Ok. I figured we might have some health check that only reported ready once the walreceiver and other infrastructure was fully up and running. If not, it's probably not worth the yak shave. Arbitrary sleeps should generally be avoided, since they're likely to introduce flake if the test runner slows down for whatever reason.

Reply:
it's launched in an async fashion, and I couldn't find a single log line that is printed on success.

There is a log line "Subscribed for broker timeline updates", but it's on the debug log level. I can change the log level if you think that's a good idea, then we can wait for that one.

Reply:
I have removed one of the sleeps without replacement, and replaced the other sleep with a wait on the "Subscribed for broker timeline updates" log line, which I put onto the info level.